File: number_of_provinces.java
class Solution {
    public int findCircleNum(int[][] isConnected) {
        
        int provinces=0;
        int N=isConnected.length;
        
        ArrayList<ArrayList<Integer>> adj=new ArrayList<>();
        for(int i=0;i<N;i++)adj.add(new ArrayList<>());
        
        for(int i=0;i<isConnected.length;i++)
        {
            for(int j=0;j<isConnected[0].length;j++){
                if(isConnected[i][j]==1&&i!=j){
                    adj.get(i).add(j);
                }
            }
        }
       
        
        int vis[]=new int[N];
        for(int i=0;i<N;i++){
            if(vis[i]==0)
            {
                provinces++;
                dfs(i,vis,adj);
            }
        }
        return provinces;
    }
    
    public void dfs(int node,int vis[],ArrayList<ArrayList<Integer>>adj){
        vis[node]=1;
        for(int it:adj.get(node)){
            if(vis[it]==0){
                dfs(it,vis,adj);
            }
        }
    }
}

File: Number_Of_Provinces.py
class Solution:
    def findCircleNum(self, isConnected: List[List[int]]) -> int:
        adj = collections.defaultdict(list)
        for i in range(len(isConnected)) :
            for j in range(len(isConnected)) :
                if isConnected[i][j]==1 and i != j :
                    adj[i].append(j)
                    adj[j].append(i)
        def dfs(i,vis) :
            vis[i]=1
            for el in (adj[i]) :
                if vis[el]==0 :
                    dfs(el,vis)
        province = 0
        vis = [0] * len(isConnected) 
        for i in range(len(isConnected)) :
            if vis[i] == 0 :
                province +=1
                dfs(i,vis)
        return province
